### PR TITLE
refactor(auth): pass the read-only gates a narrowed request view

### LINKED ISSUE #248: auth: pass the read-only gates a narrowed request view instead of *http.Request
**Context**

#243 and #247 were the same shape of defect: the authorization gates classify one thing while the wire carries another, because `AuthorizeAction` receives the whole mutable `*http.Request` and each connector decides for itself which part of it is authoritative. #246 closed that specific divergence by removing the second authority and pinning its absence with a linter.

A linter is the right tool for "nobody may touch this field". It is not the right tool for "this gate cannot see anything it should not". That is a type problem, and it is worth solving on its own merits rather than as part of a security fix.

**Proposal**

Give the read-only gates a narrowed, immutable request view instead of `*http.Request`, constructed in one place, carrying only what a gate legitimately needs: method, the URL parts, a read-only header view, and the request body as a string.

`InjectAuth` keeps the real `*http.Request`, since it mutates headers and AWS signs during it. The split is coherent: the read-only gates take the view, the mutating injection takes the request.

**What it would buy, concretely**

Two pieces of machinery exist today only because the gates are handed a live request:

- AWS reads the body during classification, then rewinds it so the signer still sees the payload (`internal/auth/aws/classifier_proto.go:69-74`):
  ```go
  body, err := io.ReadAll(req.Body)
  _ = req.Body.Close()
  ...
  req.Body = io.NopCloser(strings.NewReader(string(body)))
  ```
  This is load-bearing. `getPayloadHash` (`internal/auth/aws.go:256`) reads the body a second time during `InjectAuth` to compute the SigV4 payload hash, so dropping the rewind would hash an empty payload while sending a real one.

- GitLab avoids the problem by re-parsing the body out of the `http_request` rawArgs JSON instead of touching `req.Body` (`internal/auth/gitlab.go:431-440`), which is a correct workaround for an interface that hands over more than it should.

A view carrying `Body string` makes the rewind unnecessary and the re-parse redundant, and it removes the only place in the authorization path where a gate mutates the request it is judging.

**Scale, measured**

12 production `AuthorizeAction` implementations:

```
internal/auth/aks.go:210          internal/auth/gcp.go:114
internal/auth/aws.go:176          internal/auth/gcp/provider.go:51
internal/auth/aws/provider.go:62  internal/auth/github.go:121
internal/auth/azure.go:108        internal/auth/gitlab.go:421
internal/auth/azure/provider.go:44 internal/auth/gke.go:157
internal/auth/eks.go:159          internal/auth/kubernetes.go:482
```

Plus the request-taking helpers each one calls, the test fakes, and the test suites that construct `http.Request` values today. It is a large mechanical diff in the most safety-critical package, which is exactly why it did not belong in #246.

**Honest counterargument**

This does not remove an equality check, it relocates one into the view constructor, and the structural guarantee is partial: `InjectAuth` still receives the real request, and the SigV4 signer reads `req.Host`. So it should be judged on making the read-only gates provably pure and deleting the two workarounds above, not on being a second line of defence for #243. That job is already done by the transport rejection plus the lint pin.

Worth doing if the purity and the deletions are worth the churn. Not urgent.

### PR DESCRIPTION
**Pull request by 0xh0b0**
Closes #248.

## What & why

`AuthorizeAction` received the whole mutable `*http.Request`, so each connector
decided for itself which part of it was authoritative. That divergence is what
#243 (GCP service-claim bypass) and #247 (GitHub exposure-ceiling bypass) both
were. #246 removed the specific divergence and pinned its absence with a linter.
A linter is the right tool for "nobody may touch this field"; it is not the
right tool for "this gate cannot see anything it should not". That is a type
problem, and this is the type.

The read-only gates now take an immutable value view. `InjectAuth` keeps the
real request, since it mutates headers and AWS SigV4 signs during it.

- `ActionAuthorizer.AuthorizeAction` takes `authreq.View` (method, hostname,
  port, decoded path, escaped path, raw query, a cloned header, body string).
- `ResponseAuditor.AuditResponse` takes the smaller `authreq.AuditView` (method
  and escaped path only), built after injection so the audit still observes the
  request as sent.
- Both live in the new stdlib-only leaf `internal/auth/authreq`, because
  `internal/auth` imports its own aws/gcp/azure subpackages and a type declared
  in `internal/auth` would be an import cycle.

Two pieces of machinery existed only because the gates were handed a live
request, and both are deleted:

- AWS `classifyQueryBody` read the body during classification and rewound it so
  the SigV4 signer still saw the payload. That was a read-only gate mutating the
  request it was judging. The body now arrives as a string; `getPayloadHash` in
  `InjectAuth` is the sole remaining reader.
- GitLab re-parsed the body out of the separate `http_request` rawArgs JSON,
  because consuming `req.Body` would have broken the wire send. It authorized
  bytes from a different source than the ones sent. That re-parse, its unmarshal
  error branch, and the coverage-only test guarding it are gone.

## Shape notes

- `Hostname` and `Port` are separate and both derived only from the URL. There
  is deliberately no `Host` field: a single one would reintroduce the overloaded
  authority vocabulary behind #243 and #247.
- Both `Path` and `EscapedPath` are carried, because Azure compares them to
  reject a percent-encoded slash. Carrying one would silently disable that.
- `RawQuery` is carried unparsed with no `Query()` helper. GitLab parses
  fail-closed because Rack honors `;` as a separator and Go does not, while the
  Kubernetes and AWS gates match their Go upstreams' lenient parse. Those are
  correctly different, and one shared helper would impose a single policy on
  servers that disagree.
- The header is cloned, so a gate cannot reach the wire request through it.
  Cloning preserves key spelling, which GitLab's raw-key `Sudo` scan needs.

## What this does not close

`rawArgs` is still the full tool-call JSON passed to all six gate methods, so it
remains a second full-fidelity view of the request. Nothing in production reads
request content from it after this change (GitLab's body read was the last one),
but the narrowing is enforced by the request type only, not by everything a gate
receives. Narrowing `rawArgs` is follow-up work, not bundled here.

## Behavior

No intended behavior changes beyond the two deletions. The one candidate,
GitLab's port comparison, is equivalent: `http.NewRequestWithContext` calls
`removeEmptyPort`, so `https://host:/p` reaches the gate as `host` and both the
old `portOfAuthority(req.URL.Host)` and the new `defaultedPort(v.Port)` compute
`443`. A test pins that normalization.

## Tests

New pins: authority conflict (a request whose wire authority disagrees with its
URL, asserting the view takes the URL), header isolation, non-canonical header
key preservation, hostname normalization, body dual-observation (an authorizer
spy and the server must see the same sentinel, since a server-side assertion
alone proves wire integrity but not what the gate judged), scheme-before-auth
ordering, malformed-args-before-auth ordering, and port normalization.

Removed: the AWS body-restore assertion and its `failingReader` (they pinned a
rewind and an error branch that no longer exist), and
`TestGitLabProvider_AuthorizeAction_BadRawArgs` (its branch is deleted; the
property moved to a transport boundary test asserting malformed JSON never
reaches a gate).

Verified live against real GCP, AWS and GitHub: an authorized GCP read, a GCP
write denied client-side, an AWS SigV4 POST with a form body returning 200
(exercising the classify-then-sign path end to end), and a GitHub
secret-scanning request denied by the exposure ceiling.

## Checklist
- [x] `make check` passes locally
- [x] Tests added/updated for the change
- [x] Docs updated if behavior changed

### CODE REVIEW

File: internal/auth/authreq/authreq.go [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+		Method:      req.Method,
+		Hostname:    strings.ToLower(req.URL.Hostname()),
+		Port:        req.URL.Port(),
+		Path:        req.URL.Path,
+		EscapedPath: req.URL.EscapedPath(),
+		RawQuery:    req.URL.RawQuery,
+		Header:      header,
+		Body:        body,
+	}
+}
```
[the remark this answers is not in the document]
Comment (0xh0b0) [reply internal/auth/authreq/authreq.go:64]: Deliberate, and I'd argue a guard here would be a downgrade rather than hardening.

The suggested form returns a zero `View` when `req` or `req.URL` is nil. A zero `View` has an empty method, hostname and path, and this value is fed straight into the authorization gates. So instead of crashing on a programming error, the process would quietly ask the gates to authorize an empty request at the exact chokepoint this PR exists to narrow. A panic is the better failure mode here: loud, immediate, and impossible to mistake for a successful authorization.

The precondition is also genuinely held rather than assumed. `http.NewRequestWithContext` returns `nil, err` when the URL will not parse, and the transport checks that error, so a non-nil request always carries a non-nil URL. There is one production call site and it sits a few lines downstream of that constructor.

There is a secondary cost too: this repo enforces 100% statement coverage outside `*_shell.go`, so each added branch would need a test constructing a nil request, pinning behavior for an input that cannot occur in production.

The constructor documents the precondition explicitly, including why it is not validated: an error return would pull the projection into the fail-closed reasoning it is meant to stay out of.

File: internal/auth/authreq/authreq.go [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+// NewAuditView projects req into the AuditView handed to a response auditor.
+// Like NewView it requires a non-nil req with a non-nil URL. It is built after
+// credential injection, preserving that the audit observes the request as sent.
+func NewAuditView(req *http.Request) AuditView {
+	return AuditView{
+		Method:      req.Method,
+		EscapedPath: req.URL.EscapedPath(),
+	}
+}
```
[the remark this answers is not in the document]
Comment (0xh0b0) [reply internal/auth/authreq/authreq.go:82]: Same reasoning as on `NewView` above, and the zero-value return matters a little more here.

`AuditView{}` would carry an empty method and escaped path. Those are exactly the two values the GitHub drift audit classifies with (`RequiredLevel(v.Method, v.EscapedPath)`), so a silent zero view would produce a meaningless classification and compare it against GitHub's authoritative permissions header. That is worse than a panic, which at least surfaces the bug.

`NewAuditView` is called at one site, on the same request the transport just sent, so `req` and `req.URL` are non-nil for the same reason as in `NewView`.

**Comment by 0xh0b0** [comment]
## Live verification

Ran against real accounts on the dev VM, driving `cynative -p` through each connector's rewritten gate. 16 checks across 7 connectors.

| Connector | Read | Denial canary |
| --- | --- | --- |
| gcp | 200, projectNumber 660748312601 | `setIamPolicy` not granted by `roles/viewer` |
| aws | 200 `ListUsersResult` (SigV4 POST with a form body) | n/a |
| github | 200 repo metadata; 200 codeload HEAD | secret-scanning vs ceiling `none`; codeload POST GET/HEAD-only |
| azure | 200, 3 resource groups | non-canonical `%2F` path; provider-less PUT |
| gitlab | 200, authenticated user | `token=` in body; `Sudo` header |
| gke | 200 `PodList` | `delete pods` not in ClusterRole `view` |
| aks | 200 `NamespaceList` | n/a |

Three of these cover things the unit suite cannot reach:

- **Azure `%2F`**: denied with `non-canonical path ".../resourcegroups%2Fprobe/..."`. That control only fires if `nonCanonicalPath` still receives two distinct fields, since the decoded path carries a real `/` exactly where the escaped form carries `%2F`. Passing one field twice would leave the control dead with the whole suite still green.
- **AWS SigV4 POST with a body**: 200. The classifier read the payload from `View.Body` and `getPayloadHash` signed the same bytes. A broken rewind removal shows up here as a 403 signature mismatch, not a test failure.
- **GitLab body credential**: denied. This is the path the PR changed, from a `rawArgs` re-parse to `View.Body`. An unpopulated body would let a smuggled token through silently.

Not covered live: `eks` has no cluster in the account, and the self-managed `kubernetes` connector's kubeconfig contexts are stale, so `doctor` declines to register it. Both share the single changed line with gke and aks (`k8sauthz.Classify(v.Method, v.Path, query)` in `k8sGate.authorizeAction`), which both exercised, one on GCP credentials and one on Azure. Only their credential and endpoint resolution differs, and that is untouched here.
